csrc/executor.cpp

```diff
@@ -977,10 +977,16 @@ at::Tensor allocateOutput(
     return alloc_tensor;
   }
 
+  if (alias_info.type == AllocationType::Evaluate && !alias_info.aliased_io) {
```

Comment:
The handling of `AllocationType::Evaluate` spread at two places, which is a bit hard to follow. Wdyt about the following? 

```
optional<at::Tensor> aliased_io_tensor = std::nullopt;
if (alias_info.aliased_io != nullptr) {
  aliased_io_tensor = ...
}

switch (alias_info.type) {
  case NoAlias:
    allocate the tensor and return it;
  case InplaceUpdate:  // I thought we wanted to rename it to ReuseBuffer, but fine with a separate PR to address that. 
    return *aliased_io_tensor;
  case Evaluate:
    out_tensor = ...
    if (aliased_io_tensor.has_value()) {
      verify indeed an alias
    }
    return out_tensor;
}
```